### ISSUE TITLE
Disallow accessing federated pages by scrapers disallowed by their robots.txt

### ISSUE THREAD

**Issue by hazycora** [mastodon/mastodon#30604]
### Pitch

When a request is made to a page of a remote user’s profile, it bypasses the robots.txt of that instance. Preventing this from happening would be really nice for privacy of all users.

### Motivation

This is a significant issue as some instances don’t want to be scraped at all, particularly for the creation of AI models. For example, a site that blocks all crawling can currently still have its users scraped by mastodon.social which has a more lenient policy (while it does block the GPTBot user-agent, there are many more which aren’t blocked and it should really be up to the instances hosting that content where it should be shown, not the weakest link in terms of robots.txt policies that happens to be federating with them)

**Comment by hazycora** [comment]
here’s an example of this happening: https://social.translunar.academy/objects/81a18b5e-8bd5-4af2-a9c9-35cfa5cf002b
their robots.txt blocks all user-agents, so this shouldn’t happen if we care to enforce it on behalf of federated remote content https://social.translunar.academy/robots.txt

**Comment by renchap** [comment]
I think this would break federation entirely?

**Comment by kgourlay** [comment]
Good catch.  While there’s no way to prevent uncooperative servers from declining to share content (without defederating), mastodon should not betray robot preferences by default.

One sensible solution would be simply to maintain a policy (via e.g. `robots.txt` or `meta name="robots"` tag) to disallow crawlers on any content that is not locally controlled.  Potentially even better than this would be to mark these pages with a `link rel="canonical”` tag pointing to the canonical URL; crawlers should then respect robots.txt or other settings for the primary url.

**Comment by renchap** [comment]
There is already a `link rel="canonical"` on pages hosting content from a remote instance.

**Comment by renchap** [comment]
Also Mastodon do not serve remote public profiles to unlogged users, they are redirected to the remote instance profile URL.

**Comment by kgourlay** [comment]
Thank you @renchap I missed the rel=canonical on the page.  So is this a problem of Copilot not respecting the rel=canonical or the noindex?  Or I wonder if a crawler missed it too.  If I search for the profile mentioned by OP on mastodon.social it looks like the cached profile is displayed, with rel=canonical, both generated by javascript.  Only if I then reload the page, in order to fetch the profile url from mastodon.social directly without the javascript generation, then I get the redirect (actually an interstitial page inviting the user to redirect).  

Assuming that Copilot is using Bing as its information source, it is also possible that what it “knows” about the profile comes simply from statuses mentioning the profile on instances that allow indexing.  It is expected behavior that Bing and Goole will index statuses in this category, and thus “discover” the existence of a profile on an instance that does not allow indexing.  I don’t think there’s any way to avoid that, not to mention I’m not sure if it would be appropriate to try to block crawlers from indexing statuses that “want” to be indexed simply because they mention a user that does not want to be indexed.

Interestingly, some exploration of Bing and Google search results shows that both respect the noindex on the original server for the profile in question, but only sort of.  They declined to index the contents of the profile, but they both still index the URL and title of the page (and display a note saying effectively that indexing is blocked by robots.txt).  So once they discovered the profile URL via a public (indexable) status or website link, the search engines “know” that it exists even though they don’t index the contents of the profile.

I did notice that as far as I can tell the redirect interstitials on mastodon.social do not contain a rel=canonical or a noindex tag.  Consequently they are being indexed by Bing (and Google).  While this doesn’t really address the issue from OP, these interstitials probably should not be indexed.  They should have rel=canonical and/or noindex.

**Comment by kgourlay** [comment]
![Screen Shot 2024-06-08 at 11 32 16 AM](https://github.com/mastodon/mastodon/assets/166260365/6dc4eeb4-4fcf-4ff6-b262-5757419ea24d)

screenshot of Bing results, first the profile on the original instance (noindex) and second the redirect interstitial on an indexable instance pointing to the same profile.

**Comment by hazycora** [comment]
> Also Mastodon do not serve remote public profiles to unlogged users, they are redirected to the remote instance profile URL.
> 
> 

this isn't true. if you visit the profile through navigations, you don't redirect, whether logged in or not. with some scraping tools using full javascript environments these days, that can absolutely cause the issue I'm referring to here.

https://github.com/mastodon/mastodon/assets/72580505/b820a1e5-7b5e-4268-89df-43e2a543dc79

**Comment by ClearlyClaire** [comment]
> One sensible solution would be simply to maintain a policy (via e.g. `robots.txt` or `meta name="robots"` tag) to disallow crawlers on any content that is not locally controlled. Potentially even better than this would be to mark these pages with a `link rel="canonical”` tag pointing to the canonical URL; crawlers should then respect robots.txt or other settings for the primary url.

We do both of these things for the client-side-rendered pages for remote profiles, and do not have server-side-rendered pages for these (we put a redirect interstitial instead).

> I did notice that as far as I can tell the redirect interstitials on mastodon.social do not contain a rel=canonical or a noindex tag. Consequently they are being indexed by Bing (and Google). While this doesn’t really address the issue from OP, these interstitials probably should not be indexed. They should have rel=canonical and/or noindex.

Good catch, the redirect interstitial is missing those… I'm not completely sure a `rel=canonical` makes sense as none of the contents of the canonical page are shown on the interstitial, but I suppose it doesn't hurt.

Opened https://github.com/mastodon/mastodon/pull/30651 for this, but did not mark it as solving this issue because I think the request is different.

> if you visit the profile through navigations, you don't redirect, whether logged in or not.

That's true. We don't serve server-side-rendered pages for such contents, and an initial request directly to these pages will show a redirect interstitial instead, so well-behaved robots should not index these. The only way to completely avoid these scrappers would be to not show remote content to logged-out users at all, but this would be at the cost of a huge hit to usability.
